2022/Q2/artifacts/pos-model/PoS-model-cubic.md

```diff
@@ -0,0 +1,517 @@
+## Assumptions/simplifications
+
+- There is a single token type
+
+## Data types
+
+```go
+type Addr
+type Key
+type Epoch uint
+type VotingPower uint
+
+type Validator struct {
+  consensus_key map<Epoch, Key>
+  state map<Epoch, {inactive, candidate}>
+  total_deltas map<Epoch, amount:int>
+  voting_power map<Epoch, VotingPower>
+  reward_address Addr
+  jailed bool
+  frozen bool
```

Comment:
From what was discussed in the last meeting with Heliax folks, I understand why `frozen` is not epoched.
However, I don't think I understand why `jailed` is not epoched. Probably has to do with the fact that we don't have the "unjailing" mechanism in the model yet

Reply:
No particular reason, trying to stick to the specification: in my understanding, jailed is not epoched. We have to figure out a way to deal with epoched variables that can be updated at different offsets. Once we resolve that, we could think of merging jailed (or frozen) into the variable state.

Reply:
I agree that dealing with epoched variables that can be updated at different offsets is the biggest and most urgent of our concerns, given the ramifications it has (this is an example).